internal/keepers/service.go

```diff
@@ -132,15 +143,82 @@ func (s *onDemandUpkeepService) CheckUpkeep(ctx context.Context, keys ...types.U
 func (s *onDemandUpkeepService) start() {
 	// TODO: if this process panics, restart it
 	go s.cacheCleaner.Run(s.cache)
+	go func() {
+		if err := s.runSamplingUpkeeps(); err != nil {
+			s.logger.Fatal(err)
+		}
+	}()
 }
 
 func (s *onDemandUpkeepService) stop() {
 	close(s.stopProcs)
 	close(s.cacheCleaner.stop)
 }
 
-func (s *onDemandUpkeepService) parallelCheck(ctx context.Context, keys []types.UpkeepKey, filters ...func(types.UpkeepKey) bool) ([]*types.UpkeepResult, error) {
-	samples := newSyncedArray[*types.UpkeepResult]()
+func (s *onDemandUpkeepService) runSamplingUpkeeps() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	heads := make(chan types.BlockKey)
+	defer close(heads)
+
+	// Start the sampling upkeep process for heads
+	go func() {
+		for head := range heads {
```

Comment:
Can we just skip to the latest head here?

Reply:
Also i might be missing things, do we listen to the heads channel somewhere?

Reply:
Nope, we are going over all passed heads. It is always the latest one.

Reply:
Yes, if you look at the logic below, you could find a select statement there with the explanation. Basically, it just skips the incoming head if the current process is busy yet.